docs/02-app/02-api-reference/02-file-conventions/middleware.mdx

```diff
@@ -0,0 +1,63 @@
+---
+title: middleware.js
+description: API reference for the middleware.js  file.
+related:
+  title: Learn more about Middleware
+  links:
+    - app/building-your-application/routing/middleware
+---
+
+The middleware.ts file in Next.js is used to run code on the server before a request is finished. It provides the ability to modify the response based on the incoming request. This can include operations like URL rewriting, redirecting, altering request or response headers, or even sending a direct response.
+
+Middleware executes before the content is served from the cache and before routes are matched. It's particularly useful for implementing custom server-side functionalities like authentication, logging, or handling specific path patterns.
+
+Use the file `middleware.ts` (or .js) in the root of your project to define Middleware. For example, at the same level as pages or app, or inside src if applicable.
+
+```tsx filename="middleware.ts" switcher
+import { NextResponse } from 'next/server'
+import type { NextRequest } from 'next/server'
+
+// This function can be marked `async` if using `await` inside
+export function middleware(request: NextRequest) {
+  return NextResponse.redirect(new URL('/home', request.url))
+}
+
+export const config = {
+  matcher: '/about/:path*',
+}
+```
+
+```jsx filename="middleware.js" switcher
+import { NextResponse } from 'next/server'
+
+// This function can be marked `async` if using `await` inside
+export function middleware(request) {
+  return NextResponse.redirect(new URL('/home', request.url))
+}
+
+export const config = {
+  matcher: '/about/:path*',
+}
+```
+
+## Matcher
+
+[`matcher`](https://nextjs.org/docs/app/building-your-application/routing/middleware#matcher) allows you to filter Middleware to run on specific paths. For a single path, directly provide the path string. For multiple paths, use an array. For instance, `matcher: ['/about', '/contact']` applies Middleware to both the `/about` and `/contact` paths. Matcher supports full regular expressions, enabling complex patterns like negative lookaheads or specific character matches.
+
+## NextResponse
```

Comment:
You can add a new section called **Params** to show the arguments the default export function in the file accepts (req). Here's an example from the other file API references: https://nextjs.org/docs/app/api-reference/file-conventions/page#props

Here, for middleware, we should use `Params` instead of `Props` because it's a function, not a component. 

It might be worth noting a small quirk that usually throws people off. When importing `NextRequest` and `NextResponse`, they look similar, but one is a type and the other is a class. 

Since the `request` is passed to the default function, we give it the type, `NextRequest` . e.g. `export async function middleware(request: NextRequest)`.

`NextResponse`, on the other hand, is imported and used within the function, but not passed.